services/ngsf_analysis_service/app.py

```diff
@@ -0,0 +1,296 @@
+import ast
+from astropy.table import Table
+import io
+import os
+import functools
+import tempfile
+import base64
+import traceback
+import json
+import subprocess
+import uuid
+import zipfile
+
+import numpy as np
+import matplotlib
+import matplotlib.pyplot as plt
+import pandas as pd
+import requests
+
+from tornado.ioloop import IOLoop
+import tornado.web
+import tornado.escape
+
+from baselayer.log import make_log
+from baselayer.app.env import load_env
+
+_, cfg = load_env()
+log = make_log('ngsf_analysis_service')
+
+# we need to set the backend here to insure we
+# can render the plot headlessly
+matplotlib.use('Agg')
+rng = np.random.default_rng()
+
+default_analysis_parameters = {"fix_z": False}
+
+
+def upload_analysis_results(results, data_dict, request_timeout=60):
+    """
+    Upload the results to the webhook.
+    """
+
+    log("Uploading results to webhook")
+    if data_dict["callback_method"] != "POST":
+        log("Callback URL is not a POST URL. Skipping.")
+        return
+    url = data_dict["callback_url"]
+    try:
+        _ = requests.post(
+            url,
+            json=results,
+            timeout=request_timeout,
+        )
+    except requests.exceptions.Timeout:
+        # If we timeout here then it's precisely because
+        # we cannot write back to the SkyPortal instance.
+        # So returning something doengsf't make sense in this case.
+        # Just log it and move on...
+        log("Callback URL timedout. Skipping.")
+    except Exception as e:
+        log(f"Callback exception {e}.")
+
+
+def run_ngsf_model(data_dict):
+    """
+    Use `ngsf` to fit data to a model with name `source_name`.
+
+    For this analysis, we expect the `inputs` dictionary to have the following keys:
+       - source: the name of the model to fit to the data
+       - fix_z: whether to fix the redshift
+       - photometry: the photometry to fit to the model (in csv format)
+       - redshift: the known redshift of the object
+
+    Other analysis services may require additional keys in the `inputs` dictionary.
+    """
+    analysis_parameters = data_dict["inputs"].get("analysis_parameters", {})
+    analysis_parameters = {**default_analysis_parameters, **analysis_parameters}
```

Comment:
```suggestion
    analysis_parameters.update(default_analysis_parameters)
```

Reply:
In python 3.9 you can also do `analysis_parameters = default_analysis_parameters | analysis_parameters` though I'm not sure it is better and we are on 3.8 anyway.

Reply:
Someday we will make it to 3.9